### ISSUE TITLE
ICD11 equivalence issues

### ISSUE THREAD

**Issue by joeflack4** [monarch-initiative/icd11#4]
## Overview
While Joe was working on the [`mondo-ingest` ICD11 config PR](https://github.com/monarch-initiative/mondo-ingest/pull/412#discussion_r1493469285), he got a `robot` error on ~800 class equivalencies.

## Sub-tasks
@joeflack4 to do (1-4) and @matentzn to do (5).

- [x] **1. Fix immediate error**
In your ICD11 preprocessing code, add a ROBOT subcommand `remove --axioms equivalents`
_Edit_: Handled by https://github.com/monarch-initiative/icd11/pull/3
- [x] **2. Create a report**
Create report for all these inferred equivalents (just the relevant lines from the reasoner error output), and 
_Edit_: Handled via this [google sheet](https://docs.google.com/spreadsheets/d/1LOOdTdYWW7Ec_9QCMt-fk1UshwdyB2RxiCIxos7x3eY/edit?usp=sharing)
- [x] **3. `robot explain`**
Use [`robot explain`](https://robot.obolibrary.org/explain.html) to generate an explanation for two or three of the equivalents and post them into that issue as well.
- [x] **4. Documentation**
Document that design decision carefully in the `icd11foundation.md` docs
  - [x] Initial documentation
  - [x] Update docs with _For example X was inferred to be equivalent to Y because of the explanation Z 
(show explanation from robot)._ when Joe can get `robot explain` results.

## Additional information
The error:
```
2024-02-18 00:08:46,059 ERROR org.obolibrary.robot.ReasonOperation - Only equivalent classes that have been asserted are allowed. Inferred equivalencies are forbidden.
```

[Full log](https://drive.google.com/file/d/1ERAocYc5Myc7ASJl3wMvD_rp6MrBL0ia/view?usp=drive_link)

Related:
- #5

**Comment by joeflack4** [comment]
@matentzn I have created a report of all such problematic equivalence cases: [google sheet](https://docs.google.com/spreadsheets/d/1LOOdTdYWW7Ec_9QCMt-fk1UshwdyB2RxiCIxos7x3eY/edit?usp=sharing)

**Comment by joeflack4** [comment]
@matentzn For `robot explain`, I tried 5 different pairs that showed up in the error logs but I failed to find any explanations.

I ran like:
```
sh run.sh robot explain --input components/icd11foundation.owl --reasoner ELK \
  --axiom "http://id.who.int/icd/entity/1362736885 EquivalentTo http://id.who.int/icd/entity/2000662282" \
  --explanation icd11_problematic_equivalence1.md
```

Result:
> No explanations found.

**Comment by joeflack4** [comment]
@matentzn For this step, I could use your help:
> **4. Documentation**
> Document that design decision carefully in the `icd11foundation.md` docs

I'm not sure what to right. I have some basic notion as to why non-asserted equivalent classes would be problematic, and I understand that we are, I think as a temporary measure, doing `remove --axioms equivalents` simply so that we can continue with our work and get the build to pass without error, it doesn't seem to me that this is is what you're looking for me to document. However I certainly don't mind adding a sentence or few about this in `icd11foundation.md` if that's what you'd like.

It seems to me that such documentation would come in the future, after we'd spoken to WHO, and perhaps determined that a subset of these would be fixed and maybe a subset were going to be ignored for specific reasons. I understand documenting that. But what "design decision" would I "carefully" document at this point in time?

---

CC @twhetzel Just want you to be aware that I'm wrapping some things up in this issue.

**Comment by matentzn** [comment]
You just say something like:

Because the existing logical equivalence class axioms led to equivalence cliques (groups of distinct disease identifiers that inferred to he semantically identical) we decided to strip out all equivalence class axiom from the foundation prior to processing it in the ingest. For example X was inferred to be equivalent to Y because of the explanation Z ( show explanation from robot).

**Comment by matentzn** [comment]
The explanation command 

1) should use class labels because of Ids 
2) must he run on the original version of the ontology including the equivalent class statements.

**Comment by joeflack4** [comment]
Ah, I should've not run it on `components/icd11foundation.owl`; glad to have your eye.

I added the documentation for now, sans the _For example X was inferred to be equivalent to Y because of the explanation Z ( show explanation from robot)._ part, until I hopefully get `robot explain` to work for me, and I've added a checkbox for that in the OP.

I'm still getting no `robot explain` results. RE (1), can you explain to me why you must pass labels instead of class IDs? Is that a `robot` design decision? Seems strange. And also is it a bug that it does not throw an error when I pass it IDs instead of labels?

RE (2), I tried running on `tmp/component-download-icd11foundation.owl.owl`, as well as the release artefact I have in the [ICD11 repo](https://github.com/monarch-initiative/icd11/releases). I tried both with URIs again, just in case, as well as with labels (see command in 'Details' below). At first I got an error. I noticed that this ontology uses `skos:prefLabel`, so I changed that to `rdfs:label` for the 2 classes I was analyzing, and that solved the error, but I still got no explanation results.


<details><summary>Details</summary>
<p>

```
sh run.sh robot explain --input tmp/component-download-icd11foundation.owl.owl --reasoner ELK \
  --axiom "'Car occupant injured in collision with two- or three-wheeled motor vehicle : person on outside of vehicle injured in nontraffic accident' EquivalentTo 'Car occupant injured in collision with heavy transport vehicle or bus : person on outside of vehicle injured in nontraffic accident'" \
  --explanation icd11_problematic_equivalence1.md
```

This is the error I got before changing to `rdfs:label`:
```
Encountered 'Car occupant injured in collision with two- or three-wheeled motor vehicle : person on outside of vehicle injured in nontraffic accident' at line 1 column 1. Expected one of:
  Class name
  Object property name
  Data property name
  inv
  Functional
  inverse
  InverseFunctional
  (
  Asymmetric
  Transitive
  Irreflexive
  {
  Symmetric
  Reflexive
```

</p>
</details>

**Comment by matentzn** [comment]
Very good effort to investigate. Maybe actually using the IRI did work, but I wouldnt know. You can understand what works by opening protege and see if you can build a class expression in the class expression editor (try adding a subclass axiom). Whatever works there will work in ROBOT explain. The reason is that they wanted an "easy" way to write class expression statements, and thought people would want to use whatever they did in Protege. 

FWIW, I agree with you that the use of IRIs would be much less confusing!

The file you run it on must be pre `robot remove --axioms equivalents` else the inferences are gone! Else I do not know whats up!

**Comment by joeflack4** [comment]
Ah right, of course. How did that slip my mind. OK, I need to run this on a different file.

**Comment by joeflack4** [comment]
Alright, unfortunately I'm still getting `No explanations found.`. Here are some attempts I did, getting errors each time, until finally getting this far.

<details><summary>Running from ICD11 repo - FileNotFound</summary>
<p>

I don't know why, but it wasn't able to locate the input file.

I don't seem to be having problems running my `sh run.sh make ...`, which then does I/O just fine. yet when I run `sh run.sh robot ...` it's not able to locate the file, even though it is definitely in the same directory that this command is being run from.

```
sh run.sh robot explain --input intermediate1-unicode-cleaned.owl --reasoner ELK \
  --axiom "'Car occupant injured in collision with two- or three-wheeled motor vehicle : person on outside of vehicle injured in nontraffic accident' EquivalentTo 'Car occupant injured in collision with heavy transport vehicle or bus : person on outside of vehicle injured in nontraffic accident'" \
  --explanation icd11_problematic_equivalence1.md
org.semanticweb.owlapi.io.OWLOntologyInputSourceException: java.io.FileNotFoundException: tmp/output/intermediate1-unicode-cleaned.owl (No such file or directory)
```

I got around this problem by moving the file into `mondo-ingest` `src/ontology` and running from there.

</p>
</details> 

<details><summary>Fail on rdfs:label in .ofn</summary>
<p>

Changed `AnnotationAssertion(<http://www.w3.org/2004/02/skos/core#prefLabel>` to `AnnotationAssertion(rdfs:label`, but got:
```
Encountered 'Car occupant injured in collision with two- or three-wheeled motor vehicle : person on outside of vehicle injured in nontraffic accident' at line 1 column 1. Expected one of:
	Class name
	Object property name
	Data property name
	inv
	Functional
	inverse
	InverseFunctional
	(
	Asymmetric
	Transitive
	Irreflexive
	{
	Symmetric
	Reflexive
```

</p>
</details> 

<details><summary>Adding angle brackets</summary>
<p>

I added brackets, `AnnotationAssertion(<rdfs:label>`, but got the same error.

</p>
</details> 

<details><summary>http://www.w3.org/2000/01/rdf-schema#label</summary>
<p>

I changed to, `AnnotationAssertion(<http://www.w3.org/2000/01/rdf-schema#label>`, but got the same error.

</p>
</details> 

<details><summary>RDF/XML</summary>
<p>

I converted to RDF/XML from .ofn. It preserved my `rdfs:label` changes. But I still got ``No explanations found.`
```
sh run.sh robot explain --input icd11f.rdf.xml.owl --reasoner ELK \
  --axiom "'Car occupant injured in collision with two- or three-wheeled motor vehicle : person on outside of vehicle injured in nontraffic accident' EquivalentTo 'Car occupant injured in collision with heavy transport vehicle or bus : person on outside of vehicle injured in nontraffic accident'" \
  --explanation icd11_problematic_equivalence1.md
```

</p>
</details> 

I triple checked that I was indeed converting `skos:prefLabel` to `rdfs:label` on the correct terms; that the 2 terms IRIs in the single error message I was testing were indeed the same ones as the ones with the labels in my command, and that those were the same as what I was editing in the file. And they were the same.

I've uploaded the source file in its state before the `robot remove --axioms equivalents` step (`intermediate1-unicode-cleaned.owl`), with the `skos:prefLabel`-->`rdfs:label` edit I made, as well as the conversion of this file to RDF/XML (`icd11f.rdf.xml.owl`) to [GoogleDrive](https://drive.google.com/drive/folders/1LcoUAROrwB4O0gA1vG8RNlg2hZWCvIHw) in case you want to check my work.

**Comment by matentzn** [comment]
Good detective work

1. open intermediate1-unicode-cleaned.owl in Protege 5.6.3
3. Run Elk 0.5
4. Go to one of the classes noted in your illegal equivalence report
5. In the class description panel, you should see a line like the following (yellow highlighted) in the equivalence (not subclass!) section. Click on the `(?)` symbol in the end (you may have to click "stop search" if there are many)
6. Take a screenshot of the explanation. and post here.

**Comment by joeflack4** [comment]
Thanks for these handy instructions. Here's my report.

For my case, I used the first class mentioned in [the spreadsheet](https://docs.google.com/spreadsheets/d/1LOOdTdYWW7Ec_9QCMt-fk1UshwdyB2RxiCIxos7x3eY/edit#gid=550759389), [1362736885](http://id.who.int/icd/entity/1362736885) ([browser link](https://icd.who.int/dev11/f/en#/http%3A%2F%2Fid.who.int%2Ficd%2Fentity%2F1362736885)), _Occupant of pick-up truck or van injured in collision with other nonmotor vehicle : person on outside of vehicle injured in traffic accident_.

<details><summary>Equivalence explanation</summary>
<p>

> 'Occupant of pick-up truck or van injured in collision with other nonmotor vehicle : person on outside of vehicle injured in traffic accident.' `EquivalentTo` 'Unintentional land transport traffic event injuring an occupant of a light goods vehicle' `and` (mechanismOfInjury `some` 'Exposure to land transport on-road injury event ') `and` (modeOfTransportOfTheInjuredPerson `some` 'Pick-up truck, goods or work van, ambulance, motor home as mode of transport of person injured in transport related event') `and` (roleOfTheInjuredPerson `some` 'Person on outside of vehicle or in load space injured in transport related event')

<img width="1428" alt="1 equivalence explanation" src="https://github.com/monarch-initiative/icd11/assets/13045020/296af2e2-1e99-45a6-8dd1-a857ebf68873">

</p>
</details> 

<details><summary>Equivalent cases</summary>
<p>

I clicked the `?` (explain inference) to show this. If I'm understanding correctly, this is 4 cases of equivalent classes via this bit of reasoning:

> `EquivalentTo` 'Unintentional land transport traffic event injuring an occupant of a light goods vehicle' `and` (mechanismOfInjury `some` 'Exposure to land transport on-road injury event ') `and` (modeOfTransportOfTheInjuredPerson `some` 'Pick-up truck, goods or work van, ambulance, motor home as mode of transport of person injured in transport related event') `and` (roleOfTheInjuredPerson `some` 'Person on outside of vehicle or in load space injured in transport related event')

What's the correct way in ontology speak to refer to this bit of reasoning?

<img width="1332" alt="2 equivalent cases" src="https://github.com/monarch-initiative/icd11/assets/13045020/b9d7476e-cab6-4d37-87cb-7f25c9828e4c">

</p>
</details> 

<details><summary>Null explanation for each of the 4 cases?</summary>
<p>

For each of the 4 entries below the equivalence explanation / reasoning, I clicked the `?` (explain inference) button but got nothing. But I am imagining the equivalence explanation for each of these might be the same as what I have the expandable "Equivalence explanation" section above.

<img width="445" alt="3 no equivalence explanation for individual cases" src="https://github.com/monarch-initiative/icd11/assets/13045020/39463740-9510-407a-9173-3e0a8f339e27">

</p>
</details> 

<details><summary>Log error</summary>
<p>

I saw log errors multiple times. I think the first time I saw it was shortly after clicking "Start reasoner", and I also saw an error message pop up when clicking the `?` (explain inference) button.

Below is the full log, but here's a snippet of the log that highlights the major types of issues.
```
   INFO  17:04:25  ELK does not support positive occurrences of ObjectUnionOf or ObjectOneOf.
occurrences: ...
   INFO  17:04:25  ELK does not support querying equivalent classes and subclasses of ObjectComplementOf.
occurs in query: ...
   INFO  17:04:25  ELK does not support querying equivalent classes and superclasses of ObjectUnionOf or ObjectOneOf.
occurs in query: ...
   WARN  17:04:25  Reasoning may be incomplete! See log level INFO for more details.
   INFO  17:04:25  ELK reasoner has shut down
```

[protege.log](https://github.com/monarch-initiative/icd11/files/14470985/protege.log)

<img width="1416" alt="4 log error -idr when" src="https://github.com/monarch-initiative/icd11/assets/13045020/2fc7a8af-c64c-457e-953a-9dad0373cda6">

</p>
</details> 

<details><summary>Subclasses</summary>
<p>

FYI, in case it is useful, here are the subClassOf relationships on the class I was looking at.

<img width="1331" alt="5 subclasses" src="https://github.com/monarch-initiative/icd11/assets/13045020/4138bdd1-dc6b-4ca1-ab50-d0d542f6a4f0">

</p>
</details> 

Based on this, let me know what you'd like me to put in our documentation for this case, and if you also want me to look up any other cases.

<details><summary>Docs should say something like this?</summary>
<p>

For example http://id.who.int/icd/entity/1362736885 was inferred to be equivalent to http://id.who.int/icd/entity/2000662282 because of the explanation:
> `EquivalentTo` 'Unintentional land transport traffic event injuring an occupant of a light goods vehicle' `and` (mechanismOfInjury `some` 'Exposure to land transport on-road injury event ') `and` (modeOfTransportOfTheInjuredPerson `some` 'Pick-up truck, goods or work van, ambulance, motor home as mode of transport of person injured in transport related event') `and` (roleOfTheInjuredPerson `some` 'Person on outside of vehicle or in load space injured in transport related event')

</p>
</details>

**Comment by matentzn** [comment]
Great analysis. I think I know whats up. Can you double-check that `Occupant of pick-up truck or van injured in collision with other nonmotor vehicle : person on outside of vehicle injured in traffic accident` and all the classes that Protege shows to be equivalent to it have _identical_ equivalent class expressions? And post them here. That concludes the investigation.

As for the documentation we need to still include the exact reason for why their axioms are wrong. If the above reveals that class expressions in the axioms are indeed identical, then your documentation will just say:

> We had to remove EquivalentClass expression prior to processing as they resulted in unintended merging of foundation codes. This is based on the assumption that every single entity in the foundation is conceptually distinct, however similar.  

> What's the correct way in ontology speak to refer to this bit of reasoning?

Justification/Explanation. This corresponds to the minimal set of axioms that imply the inferences logically.

After we confirm that they use the same expression for multiple entities, we can reach out to them.

Thank you for the thorough investigation!

**Comment by joeflack4** [comment]
I took a look and I see that there are 2 different justifications for these 5 class equivalencies. Justification (1) has an additional clause/condition `and (mechanismOfInjury some 'Exposure to land transport on-road injury event ')`. Otherwise they are exactly the same. Below is a breakdown of the justifications and which classes correspond to which justification (`Eq.Id`).

### Classes
| # | ClassId | Eq.Id | Label |
|--------|--------|--------|--------|
| 1 | 1362736885 | 1 | Occupant of pick-up truck or van injured in collision with other nonmotor vehicle : person on outside of vehicle injured in traffic accident |
| 2 | 54898122 | 1 | Occupant of pick-up truck or van injured in noncollision transport accident : person on outside of vehicle injured in traffic accident |
| 3 | 2000662282 | 2 | Occupant of pick-up truck or van injured in collision with car, pick-up truck or van : person on outside of vehicle injured in traffic accident |
| 4 | 1808230283 | 1 | Occupant of pick-up truck or van injured in collision with two- or three-wheeled motor vehicle : person on outside of vehicle injured in traffic accident |
| 5 | 1279712844 | 2 | Occupant of pick-up truck or van injured in collision with heavy transport vehicle or bus : person on outside of vehicle injured in traffic accident | 

### Equivalence / Justifications
1. 'Unintentional land transport traffic event injuring an occupant of a light goods vehicle'
 `and` (mechanismOfInjury `some` 'Exposure to land transport on-road injury event ')
 `and` (modeOfTransportOfTheInjuredPerson `some` 'Pick-up truck, goods or work van, ambulance, motor home as mode of transport of person injured in transport related event')
 `and` (roleOfTheInjuredPerson `some` 'Person on outside of vehicle or in load space injured in transport related event')
2. 'Unintentional land transport traffic event injuring an occupant of a light goods vehicle'
 `and` (modeOfTransportOfTheInjuredPerson `some` 'Pick-up truck, goods or work van, ambulance, motor home as mode of transport of person injured in transport related event')
 `and` (roleOfTheInjuredPerson `some` 'Person on outside of vehicle or in load space injured in transport related event')

---

Let me know what you think in terms of next steps for investigation, docs, etc.

**Comment by matentzn** [comment]
Just a small correction for terminology:

1. Justifications are sets of axioms (like A sub B, E = C and R some B)
2. Concept expressions are statements like 
    ```
    'Unintentional land transport traffic event injuring an occupant of a light goods vehicle'
     and (mechanismOfInjury some 'Exposure to land transport on-road injury event ')
     and (modeOfTransportOfTheInjuredPerson some 'Pick-up truck, goods or work van, ambulance, motor home as mode of      transport of person injured in transport related event')
     and (roleOfTheInjuredPerson some 'Person on outside of vehicle or in load space injured in transport related event')
     ```

Your last comment answers all questions for now.

You should just document:

> "We remove all equivalent class statements as they are not unique and result in unintended node merges". For example icd11.foundation:123 (label 123) has the same exact equivalent concept expression as icd11.foundation:234 (label 234).

Then this issue can be closed. (Tag with icd11 so we can refer to it when preparing our call with them.

**Comment by joeflack4** [comment]
Ok, great. I tagged it. Note that we also have this separate follow-up issue. Trish suggested (good idea I think) that I create separately: 
- #5

Adding that documentation now.

Thanks for continuing to teach me stuff. It's frequently hard for me to ask the right questions / find the right answers when it comes to semantic web, etc.

I think (2) is slightly different than this [other example of concept expressions](https://www.inf.ed.ac.uk/teaching/courses/sws/lectures/SWS-10-DescriptionLogicsOWL.pdf) I found online:
> Parent ≡ “Persons who have (amongst other things) some children”_
> (Person ⨅ ∃ hasChild.⊤)

The difference being that in our example, _Unintentional land transport traffic event injuring an occupant of a light goods vehicle_ is followed by the `and` operator, rather than on operator like equivalence `≡` or intersection `⨅` of restriction(s).

Also I don't know if I'd say for (1) that justification is 'just' a set of axioms. In any case, my knowledge progresses.
